### ISSUE TITLE
Referencing archives

### ISSUE THREAD

**Issue by davidcassany** [containers/image#460]
I have noted that at least for `docker-archive` and I pretty sure it also happens for other `archive` types that src and dest reference strings expect something like:

`<transport>:<path>[:<reference>]`

The problem I have is that the _path_ can potentially contain colons `:`, when this happens _skopeo copy_ calls fail parsing the source image reference in a wrong way. I believe it would make sense to allow paths with a wider range of characters for archive paths, ideally whatever is supported by the underlying platform.

I could try to provide a PR solving it for `docker-archive`, probably by distinguishing the source parser (where reference should not be there) from the destination parser (where reference is required) or by trying to stat `<path>` before assuming any present colons are a field separator.

Would it make sense to you?

**Comment by mtrmac** [comment]
Thanks.

This has been discussed earlier in https://github.com/projectatomic/libpod/issues/489 .

Basically, we are stuck with having _some_ string representation, and a string representation means in-line metacharacters.

> distinguishing the source parser (where reference should not be there) from the destination parser (where reference is required)

That does not fix destinations; and whatever solution, if any, designed for destinations, would not apply to sources.  And of course it is _highly_ desirable that sources and destinations accept the same format.

> by trying to stat `<path>` before assuming any present colons are a field separator.

It’s brittle if the same string means different things in different situations.

---

As discussed in https://github.com/projectatomic/libpod/issues/489 , it’s by far simplest to ignore the problem. If that is not acceptable, we can design some escaping mechanism (“`::` means a single colon”? “`\:` means a colon, `\\` means a `\`”?) — but that will require _every single caller_ to correctly escape the command-line inputs, which is tricky to do in shell scripts, and, sadly, unlikely to happen in many of them; most people won’t read the documentation that carefully, if at all.

**Comment by davidcassany** [comment]
@mtrmac thanks for asking and linking the discussion from podman. Sorry I only checked this repo and skopeo repo for similar issues, I did not check other dependent tools.

> That does not fix destinations; and whatever solution, if any, designed for destinations, would not apply to sources.

Well the idea here was to force `<transport>:<path>` for sources, where only the first colons are separators, and force `<transport>:<path>:<container>:<tag>` for destinations, where first colons (from left) separate transport and the two first appearances (from right) separate the image reference. But I agree this complex and probably only applicable for `docker-archive`.

> And of course it is highly desirable that sources and destinations accept the same format.

Fully agree.

> It’s brittle if the same string means different things in different situations.

Yes, I also dislike this idea.
Well, that's why I asked before attempting any patch.

The escaping mechanism suggested in projectatomic/libpod#489 sounds quite nice, haven't thought about that. I understand your concerns regarding this option as it should be handled in many callers and probably require some extensive refactoring, but IMHO sounds reasonable.

Also, form a wishful thinking perspective, I'd love to see some day image referencing following URI standard, so solid parsing could be applied to extract scheme, path, domain, fragments, parameters, etc. I know this is unlikely to happen given how this is all designed now, but dreaming is free :smile: 

Thanks

**Comment by mtrmac** [comment]
> @mtrmac thanks for asking and linking the discussion from podman. Sorry I only checked this repo and skopeo repo for similar issues, I did not check other dependent tools.

No problem; this might have been mentioned in here somewhere as well, but the podman issue is the one I remembered, and it contains the clearest examples of the issues, specifically for `docker-archive:`.

> The escaping mechanism suggested in projectatomic/libpod#489 sounds quite nice, haven't thought about that. I understand your concerns regarding this option as it should be handled in many callers and probably require some extensive refactoring, but IMHO sounds reasonable.

The code aspect does not worry me _that_ much; all the parsing and formatting is consolidated in a single file for each transport, with required good test code coverage (because of its interactions with signature verification).

The UX and callers escaping/not escaping correctly is a bigger problem — but maybe we really can’t get away with forbidding `:` in paths.  I don’t know. What is your actual use case which requires colons? And @rhatdan @baude @mheon WDYT?

> Also, form a wishful thinking perspective, I'd love to see some day image referencing following URI standard, so solid parsing could be applied to extract scheme, path, domain, fragments, parameters, etc.

- (As far as the _generic_ containers/image code cares, the only thing that is mandatory is the _transport_`:` prefix. A particular transport could decide to use an URI-like format.)
- Using URIs doesn’t really help with the metacharacter problem, because URIs also have in-line metacharacters (only possibly different ones, and actually many more than the existing transports do), and we would still need escaping.
- The URI percent-escaping is absolutely _horrible_ for human readers/writers (and not exactly trivial for shell scripts), and image names are frequently enough written in CLIs that forcing humans to write percent escapes seems hard to accept. (If we gave up on human reading/writing, we can just as well go full JSON, or a full binary-representation-of-Golang-structure, or protobuf, or something like that, and get actual type safety.)

**Comment by vrothberg** [comment]
> The UX and callers escaping/not escaping correctly is a bigger problem — but maybe we really can’t get away with forbidding : in paths. I don’t know. What is your actual use case which requires colons? And @rhatdan @baude @mheon WDYT?

@mtrmac This issue came up today internally again. https://openbuildservice.org is building container images using Kiwi+umoci but also using Dockerfiles+docker. The build service team is currently evaluating podman (also for building) and hit the issue again.

> What is your actual use case which requires colons?

I can think of two concrete use cases:
 1) Mapping the `image:tag` to the file system.
 2) The open build service uses ":" to split parent from child projects. Having a tarball with ":" would allow a direct mapping from the build project to the build artifact.

ping @baude @rhatdan @umohnani8 @mheon @TomSweeneyRedHat

**Comment by mtrmac** [comment]
Oh well, if we have, to, we have to. (Others WDYT?)

Step 0 here would be to add a `c/image/docker/archive.NewReference(string, reference.NamedTagged)`, so that API callers can submit paths and references which are not subject to unescaping; and so that they can drop their specific knowledge that `:` is forbidden. (And `transports.ImageName(archive.NewReference())` would do the escaping to create string representations correctly, by definition). I need that for podman cleanups as well, though I’m not quite sure when I will get to it.

Then we can bikeshed about what the escaping format should be. The only hard requirement for me is that the format **must** be unambiguous.

As for soft preferences, it should be, if at all possible independent of filesystem contents, and I can’t quite imagine humans writing percent-escapes.  Beyond that, anything goes?

(Another possible consideration: `:` is a natural component of Windows paths; compare `c/image/oci/internal`.)

**Comment by mtrmac** [comment]
> so that API callers can submit paths and references which are not subject to unescaping

… and so that API callers, after converting to `NewReference`, are not affected by whatever string representation we end up using.

**Comment by rhatdan** [comment]
I am fine with this change.

**Comment by rhatdan** [comment]
@davidcassany  @mtrmac @vrothberg What is the state of this issue?

**Comment by vrothberg** [comment]
Nothing has changed in the code but I believe that OBS found a solution in the meantime. @davidcassany, do you know about OBS?

**Comment by davidcassany** [comment]
@vrothberg, @rhatdan we are certainly building images with OBS thus we are not facing this issue at moment. I had to review once again the related code in OBS and in KIWI and drain my memory then I just realized we did not make any action to solve it, it got casually _solved_ by using compressed tarballs for images in OBS. Using compressed tarballs causes KIWI to decompress images before calling `skopeo`, the uncompressed results are stored in a temporary folder that is not including any colons in its name. 

Form our side I guess this issue could be closed and if it raises again in our side we still have the chance to workaround in OBS by changing the default paths used in the build environment. I presume this is what OBS is doing for podman builds.
